open-craft/edx-platform
# Learner dashboard: Populate list of pending reviews with data from adaptive learning service

This PR follows up on #47, which added basic functionality for displaying a list of pending revisions on a learner's dashboard. It adds logic for displaying actual revisions (instead of a list of links to hard-coded blocks).

This PR also updates the URL for fetching pending reviews to point to the new API that Domoscio introduced a few days ago.

cf. [OC-2154](https://tasks.opencraft.com/browse/OC-2154)

**Screenshots**

![dashboard](https://cloud.githubusercontent.com/assets/961441/21700612/90df8386-d3a1-11e6-8ea6-3727fed2069b.png)

**Basic setup**

1. Follow instructions on the [epic](https://tasks.opencraft.com/browse/OC-2127) to set up a FUN box.

2. Switch to this branch (`display-revisions`) and install the new block via

    ```sh
    pip install -r requirements/edx/local.txt
    ```

3. Create a new content library and add a few problems to it.

4. Create a new course and add `"adaptive_library_content"` to advanced modules.

5. Add an "Adaptive Content Block" that references the library created previously to a unit (optionally modifying the number of components to display to each student), and publish the unit. Do **not** access the unit in the LMS yet.

6. To be able to work with the Domoscio API, you'll need to specify a few course-level settings: In Studio, navigate to the Advanced Settings page for the course ("Paramètres" > "Paramètres avancés") and replace the default value for the "Adaptive Learning Configuration" setting with the following:

    ```json
    "url": "<see comment on internal ticket>",
    "instance_id": <see comment on internal ticket>,
    "access_token": "<see comment on internal ticket>",
    "api_version": "v1"
    ```

7. In a venv that has the `requests` library installed, create a knowledge node that corresponds to the unit you created earlier (`block_id` needs to be a 32-character hash), and review the results:

    ```python
    import requests, json
    response = requests.post("<url>/v1/instances/<instance_id>/knowledge_nodes", headers={"Authorization": "Token token=<access_token>"}, data={"knowledge_graph_id": <see comment on internal ticket>, "name": "Unit node", "uid": "<block_id of unit>"})
    json.loads(response.content)
    ```

    Then create a knowledge node for each of the problems that belong to the library you created earlier. Note that problems belonging to a content library are assigned unique block IDs for each Adaptive Content Block that references them; you'll need to use these context-specific IDs when creating corresponding knowledge nodes. Block IDs for problems are 20-character hashes.

**Test instructions**

1. Review the list of links between students and units ("knowledge node student" objects) on the Domoscio test instance:

    ```python
    response = requests.get("<url>/v1/instances/<instance_id>/knowledge_node_students", headers={"Authorization": "Token token=<access_token>"})
    json.loads(response.content)    
    ```
2. Access the unit that contains the ACB in the LMS.

3. Review the list of "knowledge node student" objects as described above. Observe that it contains new "knowledge node student" objects, one for each knowledge node you created earlier.

4. Identify the "knowledge node student" object that links the current user to the *unit* containing the ACB. Make a note of its `"id"`. Also make a note of the `"knowledge_node_uid"` of one of the "knowledge node student" objects that link the current user to a *problem* belonging to the ACB.

5. Update the `"next_review_at"` field of the "knowledge node student" object, to force a review to become available for this object:

    ```python
    from datetime import datetime
    d = datetime(2016, 12, 24)
    # Replace <knowledge_node_student_id> with the "id" you wrote down in the previous step:
    response = requests.put("<url>/<instance_id>/knowledge_node_students/<knowledge_node_student_id>", headers={"Authorization": "Token token=<token>"}, data={"next_review_at": d.isoformat()})
    # Status should be 204
    response.status_code  
    ```

6. Temporarily change [utils.py#L35](https://github.com/open-craft/edx-platform/blob/display-revisions/common/djangoapps/adaptive_learning/utils.py#L35) to:

    ```python
    # Replace <knowledge_node_uid> with "knowledge_node_uid" you wrote down two steps back
    pending_review['review_question_uid'] or '<knowledge_node_uid>': pending_review['next_review_at']
    ```

    **Explanation**: This step is necessary because the data on the Domoscio test instance that we've been using is missing links between units and problems belonging to ACBs. (Domoscio staff are creating these links manually for the production instance; our code does not have to create them. We can't perform this step on the test instance ourselves because Domoscio has not disclosed the necessary details.) Because of this, any results returned by the API for fetching pending reviews will have the `"review_question_uid"` property set to `None`. As a result, even though we forced a review to become available in the previous step, the (unmodified) code will not find a matching block for that review in the modulestore.

7. Finally, navigate to the dashboard by clicking the name of the current user in the top right corner. Observe that the review that you forced to become available earlier is listed on the right. The name and due date for the review should be equal to the `display_name` of the corresponding problem and the date you forced for the review earlier, respectively. Clicking on the review should take you to the unit that contains the parent ACB. (Note that the ACB won't show the problem to review; this is, again, due to the fact that the data on the Domoscio test instance is missing links that will only be available on the production instance.)

**Relevant tests**

The commands for running relevant tests are:

```sh
paver test_lib -l common/lib/xmodule/xmodule/tests/test_library_content.py
paver test_lib -l common/lib/xmodule/xmodule/tests/test_util_adaptive_learning.py
paver test_system -t common/djangoapps/adaptive_learning/tests/test_views.py --fasttest
paver test_system -t common/djangoapps/adaptive_learning/tests/test_utils.py --fasttest
```

**Reviewers**
